projects/extension/sql/idempotent/905-text-to-sql.sql

```diff
@@ -0,0 +1,510 @@
+as $func$
+declare
+    _nspname pg_catalog.name;
+    _relname pg_catalog.name;
+    _columns pg_catalog.text[];
+    _constraints pg_catalog.text[];
+    _indexes pg_catalog.text;
+    _ddl pg_catalog.text;
+begin
+    -- names
+    select
+      n.nspname
+    , k.relname
+    into strict
+      _nspname
+    , _relname
+    from pg_catalog.pg_class k
+    inner join pg_catalog.pg_namespace n
+    on (k.relnamespace operator(pg_catalog.=) n.oid)
+    where k.oid operator(pg_catalog.=) objid
+    ;
+
+    -- columns
+    select pg_catalog.array_agg(x.txt order by x.attnum)
+    into strict _columns
+    from
+    (
+        select pg_catalog.concat_ws
+        ( ' '
+        , a.attname
+        , pg_catalog.format_type(a.atttypid, a.atttypmod)
+        , case when a.attnotnull then 'NOT NULL' else '' end
+        , case
+            when a.atthasdef
+                then pg_catalog.pg_get_expr(d.adbin, d.adrelid)
+            when a.attidentity operator(pg_catalog.=) 'd'
+                then 'GENERATED BY DEFAULT AS IDENTITY'
+            when a.attidentity operator(pg_catalog.=) 'a'
+                then 'GENERATED ALWAYS AS IDENTITY'
+            when a.attgenerated operator(pg_catalog.=) 's'
+                then pg_catalog.format('GENERATED ALWAYS AS (%s) STORED', pg_catalog.pg_get_expr(d.adbin, d.adrelid))
+            else ''
+          end
+        ) as txt
+        , a.attnum
+        from pg_catalog.pg_attribute a
+        left outer join pg_catalog.pg_attrdef d
+        on (a.attrelid operator(pg_catalog.=) d.adrelid and a.attnum operator(pg_catalog.=) d.adnum)
+        where a.attrelid operator(pg_catalog.=) objid
+        and a.attnum operator(pg_catalog.>) 0
+        and not a.attisdropped
+    ) x;
+
+    -- constraints
+    select pg_catalog.array_agg(pg_catalog.pg_get_constraintdef(k.oid, true) order by k.conname)
+    into _constraints
+    from pg_catalog.pg_constraint k
+    where k.conrelid operator(pg_catalog.=) objid
+    ;
+
+    -- indexes
+    select pg_catalog.string_agg(pg_catalog.pg_get_indexdef(i.indexrelid, 0, true), E';\n')
+    into strict _indexes
+    from pg_catalog.pg_index i
+    where i.indrelid operator(pg_catalog.=) objid
+    ;
+
+    -- ddl
+    select pg_catalog.format(E'CREATE TABLE %I.%I\n( ', _nspname, _relname)
+    operator(pg_catalog.||)
+    pg_catalog.string_agg(x.line, E'\n, ')
+    operator(pg_catalog.||) E'\n);\n'
+    operator(pg_catalog.||) _indexes
+    into strict _ddl
+    from
+    (
+        select * from pg_catalog.unnest(_columns) line
+        union all
+        select * from pg_catalog.unnest(_constraints) line
+    ) x
+    ;
+
+    return _ddl;
+end
+$func$ language plpgsql stable security invoker
+set search_path to pg_catalog, pg_temp
+;
+
+-------------------------------------------------------------------------------
+-- _text_to_sql_prompt
+create or replace function ai._text_to_sql_prompt
```

Comment:
we should think about how to make these functions more composable and customizable. This feels too canned for me.